### PR TITLE
feat: add React DevTools

### PR DESCRIPTION
**Pull request by FedeIlLeone**
- Setting to enable React DevTools.
- Advanced settings have been moved to a new dedicated category.

Closes: #253

### CODE REVIEW

File: src/main/ipc/react-devtools.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import fetch from "node-fetch";
+import { join } from "path";
+import { CONFIG_PATHS } from "src/util.mjs";
+import { RepluggedIpcChannels } from "../../types";
+
+const OUTPUT_PATH = join(CONFIG_PATHS["react-devtools"]);
+
+// https://github.com/facebook/react/issues/25843#issuecomment-1406766561
+const EXTENSION_URL =
+  "https://github.com/mondaychen/react/raw/017f120369d80a21c0e122106bd7ca1faa48b8ee/packages/react-devtools-extensions/ReactDevTools.zip";
```
Comment (asportnoy) [inline src/main/ipc/react-devtools.ts:12]: Per internal discussion, we want to move this to some source we control, either forking this to our GH org or hosting it directly on the server.

File: src/main/ipc/react-devtools.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+ipcMain.handle(RepluggedIpcChannels.DOWNLOAD_REACT_DEVTOOLS, async () => {
+  const arrayBuffer = await fetch(EXTENSION_URL)
```
Comment (asportnoy) [inline src/main/ipc/react-devtools.ts:15]: Can we check if it's already there before downloading it again? Also one we should consider (which ig could be done later) is figure out how to handle updates to RDT. Maybe a setting key for the version and if that's older than whatever version is in a certain const then we re-download when Discord is launched?

File: src/renderer/coremods/settings/pages/General.tsx [hunk 3]
```diff
@@ -86,42 +80,84 @@ export const General = (): React.ReactElement => {
...
+              .confirm({
+                title: Messages.REPLUGGED_SETTINGS_REACT_DEVTOOLS,
+                body: Messages.REPLUGGED_SETTINGS_REACT_DEVTOOLS_DESC.format(),
+                confirmText: Messages.REPLUGGED_RELOAD,
+                confirmColor: Button.Colors.RED,
+              })
+              .then((answer) => {
+                if (answer) {
+                  if (value) {
+                    void RepluggedNative.reactDevTools.downloadExtension().then(() => {
```
Comment (asportnoy) [inline src/renderer/coremods/settings/pages/General.tsx:137]: Add error handling for download failed. Maybe just a toast?

File: src/renderer/coremods/settings/pages/General.tsx [hunk 4]
```diff
@@ -86,42 +80,84 @@ export const General = (): React.ReactElement => {
...
+                      rdtOnChange(value);
+                      setTimeout(() => window.DiscordNative.app.relaunch(), 250);
```
Comment (asportnoy) [inline src/renderer/coremods/settings/pages/General.tsx:139]: Let's create functions here called reload and relaunch (with the timeout built in) and then replace the calls in this file.

File: src/types/settings.ts [hunk 5]
```diff
@@ -12,6 +12,7 @@ export type GeneralSettings = {
   badges: boolean;
   autoApplyQuickCss: boolean;
   showWelcomeNoticeOnOpen: boolean;
+  reactDevTools: boolean;
 };
 
 export const defaultSettings: Partial<GeneralSettings> = {
```
Comment (asportnoy) [inline src/types/settings.ts:18]: Unrelated to this PR but the type here should be using `satisfies`:

```ts
export const defaultSettings = {
  badges: true,
  autoApplyQuickCss: false,
  showWelcomeNoticeOnOpen: true,
  reactDevTools: false,
} satisfies Partial<GeneralSettings>;
```
GH won't let me suggest here but can you fix it in this branch?

**Review by asportnoy — CHANGES_REQUESTED** [review]

File: src/main/ipc/react-devtools.ts (hunk 2 above)
Comment (FedeIlLeone) [reply src/main/ipc/react-devtools.ts:15]: > Can we check if it's already there before downloading it again?
What if the extension breaks, I don't know, the user deletes some file or changes something making it not loading. A redownload would overwrite any files, solving this possible issue. How would I check if the extension is downloaded or not? Should I save the zip individually, and check if it's there?

> Also one we should consider (which ig could be done later) is figure out how to handle updates to RDT. Maybe a setting key for the version and if that's older than whatever version is in a certain const then we re-download when Discord is launched?
Certainly later, if we will consider forking the modified RDT.

File: src/main/ipc/react-devtools.ts [hunk 6]
```diff
@@ -1,23 +1,36 @@
...
-  const arrayBuffer = await fetch(EXTENSION_URL)
-    .then((res) => res.arrayBuffer())
-    .catch(() => {
-      throw new Error("Could not download React DevTools");
-    });
-  const buffer = Buffer.from(new Uint8Array(arrayBuffer));
+  const apiUrl =
+    ((await readTransaction("dev.replugged.Settings", (settings) => settings.get("apiUrl"))) as
+      | string
+      | undefined) ?? "https://replugged.dev";
```
Comment (asportnoy) [inline src/main/ipc/react-devtools.ts:17]: Can we set up some ipc util with this? Since we're using it in store too

File: src/renderer/coremods/settings/pages/General.tsx [hunk 7]
```diff
@@ -134,12 +142,21 @@ export const General = (): React.ReactElement => {
...
-                      setTimeout(() => window.DiscordNative.app.relaunch(), 250);
-                    });
+                    void RepluggedNative.reactDevTools
+                      .downloadExtension()
+                      .then(() => {
+                        rdtOnChange(value);
+                        relaunch();
+                      })
+                      .catch(() => {
+                        void toast.toast(
```
Comment (asportnoy) [inline src/renderer/coremods/settings/pages/General.tsx:152]: Not a promise so `void` shouldn't be needed
```suggestion
                        toast.toast(
```

File: src/renderer/coremods/settings/pages/General.tsx [hunk 8]
```diff
@@ -86,42 +88,93 @@ export const General = (): React.ReactElement => {
...
+        <SwitchItem
+          value={rdtValue}
+          onChange={(value) => {
+            void modal
+              .confirm({
+                title: Messages.REPLUGGED_SETTINGS_REACT_DEVTOOLS,
+                body: Messages.REPLUGGED_SETTINGS_REACT_DEVTOOLS_DESC.format(),
+                confirmText: Messages.REPLUGGED_RELOAD,
+                confirmColor: Button.Colors.RED,
+              })
```
Comment (asportnoy) [inline src/renderer/coremods/settings/pages/General.tsx:141]: Can we download first and then show the modal?

**Review by asportnoy — CHANGES_REQUESTED** [review]

**Comment by lexisother** [comment]
Is this done? If there are no more comments to be made I am dismissing your review as stale. @asportnoy

**Review by lexisother — APPROVED** [review]

File: src/renderer/coremods/settings/pages/General.tsx [hunk 9]
```diff
@@ -110,56 +124,31 @@ export const General = (): React.ReactElement => {
...
-                confirmColor: Button.Colors.RED,
-              })
-              .then((answer) => {
-                if (answer) {
-                  expOnChange(value);
-                  reload();
-                }
-              });
-          }}
+          onChange={(value) => restartModal(false, () => expOnChange(value))}
```
Comment (asportnoy) [inline src/renderer/coremods/settings/pages/General.tsx:127]: This should change the setting even if you choose to not restart yet

File: src/renderer/coremods/settings/pages/General.tsx [hunk 10]
```diff
@@ -110,56 +124,31 @@ export const General = (): React.ReactElement => {
...
-                    rdtOnChange(value);
-                    relaunch();
-                  }
-                }
-              });
+            if (value) {
+              void RepluggedNative.reactDevTools
+                .downloadExtension()
+                .then(() => {
+                  rdtOnChange(value);
```
Comment (asportnoy) [inline src/renderer/coremods/settings/pages/General.tsx:139]: Do the onChange before downloading so the switch doesn't flicker, but revert the change in the catch block.

**Review by asportnoy — CHANGES_REQUESTED** [review]
Last two things hopefully

**Review by asportnoy — APPROVED** [review]
image
Will merge once Fede gives the ok
